app/models/transformation_mapping_item.rb

```diff
@@ -24,4 +31,36 @@ def destination_cluster
       errors.add(:destination, "EMS type of destination cluster or cloud tenant must be in: #{destination_types}")
     end
   end
+
+  def source_datastore
+    source_storage = source
+
+    cluster_storages = source.hosts # Get hosts using this source storage
+                             .collect(&:ems_cluster) # How many clusters does each host has
```

Comment:
Just a nit, the comment should read: "How many clusters does each host have"

Reply:
What are you trying to check here? It looks like you're making sure the storage is in a set which has to include itself?  If the storage wasn't in the host's cluster then it wouldn't have that host in the first place.  Feels circular

Reply:
Yes this is how I felt but the `TransformationMappingItem` model did not have any other way of getting to the cluster this `TransformationMappingItem` was associated with.   The question is, is it okay for the first attempt at `TransformationMappingItem` validation and comeback to it in future as it may require changing this and other associated models.

Reply:
Okay so you're trying to make sure that the cluster they selected contains the storage they selected? That makes sense but that's not what is happening here.

You probably want to get the cluster they selected by going to the parent transformation mapping and finding the cluster mapping item right?

Reply:
Yes, I check the parent `TranspformationMapping`, its attributes `ServiceResource` and `ServiceTemplate`but did not find it.

Reply:
I meant to say, Yes, I **checked** the parent TranspformationMapping, its attributes `ServiceResource` and `ServiceTemplate` but did not find any reference to `cluster`.  This means it requires changing the model and schema which broadens its scope and needs reevaluation to get it fixed in future.

Reply:
Done